pkg/client/apiutil/lazyrestmapper.go

```diff
@@ -0,0 +1,260 @@
+			return res, err
+		}
+		res, err = m.mapper.RESTMappings(gk, versions...)
+	}
+	return res, err
+}
+
+// ResourceSingularizer implements Mapper.ResourceSingularizer
+func (m *LazyRESTMapper) ResourceSingularizer(resource string) (string, error) {
+	res, err := m.mapper.ResourceSingularizer(resource)
+	if meta.IsNoMatchError(err) {
+		res, err = m.mapper.ResourceSingularizer(resource)
+	}
+	return res, err
+}
+
+// fetchGroupVersionResources uses the discovery client to fetch the resources for the specified groups in parallel.
+// Mainly replicates the same named function from the client-go internals aside from the changed `apiGroups` argument type (uses slice instead of APIGroupList).
+// ref: https://github.com/kubernetes/kubernetes/blob/a84d877310ba5cf9237c8e8e3218229c202d3a1e/staging/src/k8s.io/client-go/discovery/discovery_client.go#L506
+func fetchGroupVersionResources(d discovery.DiscoveryInterface, apiGroups []*metav1.APIGroup) (map[schema.GroupVersion]*metav1.APIResourceList, map[schema.GroupVersion]error) {
+	groupVersionResources := make(map[schema.GroupVersion]*metav1.APIResourceList)
+	failedGroups := make(map[schema.GroupVersion]error)
+
+	wg := &sync.WaitGroup{}
+	resultLock := &sync.Mutex{}
+	for _, apiGroup := range apiGroups {
+		for _, version := range apiGroup.Versions {
+			groupVersion := schema.GroupVersion{Group: apiGroup.Name, Version: version.Version}
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				defer utilruntime.HandleCrash()
+
+				apiResourceList, err := d.ServerResourcesForGroupVersion(groupVersion.String())
+
+				// lock to record results
+				resultLock.Lock()
+				defer resultLock.Unlock()
+
+				if err != nil {
+					// TODO: maybe restrict this to NotFound errors
+					failedGroups[groupVersion] = err
+				}
+				if apiResourceList != nil {
+					// even in case of error, some fallback might have been returned
+					groupVersionResources[groupVersion] = apiResourceList
+				}
+			}()
+		}
+	}
+	wg.Wait()
+
+	return groupVersionResources, failedGroups
+}
+
+// filteredServerGroupsAndResources returns the supported resources for groups filtered by passed predicate and versions.
+// Mainly replicate ServerGroupsAndResources function from the client-go. The difference is that this function takes
+// a function of the GroupFilterPredicate type as an argument for filtering out unwanted groups.
+// ref: https://github.com/kubernetes/kubernetes/blob/a84d877310ba5cf9237c8e8e3218229c202d3a1e/staging/src/k8s.io/client-go/discovery/discovery_client.go#L383
+func (m *LazyRESTMapper) filteredServerGroupsAndResources(d discovery.DiscoveryInterface) ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
+	sgs, err := d.ServerGroups()
+	if sgs == nil {
+		return nil, nil, err
+	}
+	resultGroups := []*metav1.APIGroup{}
+	for i := range sgs.Groups {
+		if m.isGroupKnown((&sgs.Groups[i]).Name) {
+			resultGroups = append(resultGroups, &sgs.Groups[i])
+		}
+	}
```

Comment:
Every time we look for a new group, we are also going to discover the old groups as well, is that desirable? Or should we convert the logic to merge the old results with the newly requested group? In that case, if you see an object that isn't mapped, but in a known group, that group will become the "new group" and we will refresh it, but leave the others alone, which will minimise the API calls we make